모임 필터링 조회 무한루프로 인한 커넥션 풀 고갈 현상 수정

## 요구 사항
- 현재 모임 `참여 가능` 조건으로 조회 시 응답을 해주지 못하고 있음 
- 클라이언트는 서버로부터 응답을 받지 못해 타임 아웃이 발생함
- 모종의 이유로 `참여 가능` 모임 조회 시 무한 루프 도는 것으로 추정
- 해당 요청을 계속해서 보내면 커넥션이 반납되지 않아 커넥션 풀이 고갈되고, 다른 모든 요청 처리가 불가능해지는 상황이 발생함

```java
Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30008ms (total=10, active=10, idle=0, waiting=0)
        at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:686)
        at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:179)
        at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:144)
        at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:127)
        at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
        at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:46)
        at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:113)
        ... 66 common frames omitted
```

--- a/backend/src/main/java/com/happy/friendogly/club/service/ClubQueryService.java
+++ b/backend/src/main/java/com/happy/friendogly/club/service/ClubQueryService.java
@@ -60,8 +60,8 @@ public class ClubQueryService {
         addFilteredResults(request, filteredClubs, result);
 
         while (result.size() < request.pageSize() && clubSlice.hasNext()) {
-            lastFoundCreatedAt = updateLastFoundCreatedAt(lastFoundCreatedAt, result);
-            lastFoundId = updateLastFoundId(lastFoundId, result);
+            lastFoundCreatedAt = updateLastFoundCreatedAt(clubSlice, result);
+            lastFoundId = updateLastFoundId(clubSlice, result);
 
             clubSlice = fetchClubSlice(request, lastFoundCreatedAt, lastFoundId);
             filteredClubs = filterClubs(clubSlice, filterCondition, member, pets);
@@ -118,16 +118,16 @@ public class ClubQueryService {
         }
     }
 
-    private LocalDateTime updateLastFoundCreatedAt(LocalDateTime lastFoundCreatedAt, List<Club> result) {
+    private LocalDateTime updateLastFoundCreatedAt(Slice<Club> clubSlice, List<Club> result) {
         if (result.isEmpty()) {
-            return lastFoundCreatedAt;
+            return clubSlice.getContent().get(clubSlice.getSize() - 1).getCreatedAt();
         }
         return result.get(result.size() - 1).getCreatedAt();
     }
 
-    private Long updateLastFoundId(Long lastFoundId, List<Club> result) {
+    private Long updateLastFoundId(Slice<Club> clubSlice, List<Club> result) {
         if (result.isEmpty()) {
-            return lastFoundId;
+            return clubSlice.getContent().get(clubSlice.getSize() - 1).getId();
         }
         return result.get(result.size() - 1).getId();
     }

--- a/backend/src/test/java/com/happy/friendogly/club/service/ClubQueryServiceTest.java
+++ b/backend/src/test/java/com/happy/friendogly/club/service/ClubQueryServiceTest.java
@@ -17,9 +17,13 @@ import com.happy.friendogly.pet.domain.Gender;
 import com.happy.friendogly.pet.domain.Pet;
 import com.happy.friendogly.pet.domain.SizeType;
 import jakarta.transaction.Transactional;
+import java.time.LocalDate;
 import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
@@ -137,6 +141,49 @@ class ClubQueryServiceTest extends ClubServiceTest {
         FindClubPageByFilterResponse responses = clubQueryService.findByFilter(savedMember.getId(), request);
 
         assertThat(responses.content().isEmpty()).isTrue();
+    }
+
+    @DisplayName("첫 페이지에 필터링 조건에 맞는 모임이 존재하지 않는 경우에도 다음 페이지를 읽어온다.")
+    @Test
+    void findSearching_NoResultOnFirstPage_MoveToNextPage() {
+        List<Club> clubs = new ArrayList<>();
+        clubs.add(createSavedClub(    // member가 참여 불가능한 모임
+                savedMember,
+                List.of(savedPet),
+                Set.of(Gender.FEMALE),
+                Set.of(SizeType.SMALL)
+        ));
+        clubs.add(createSavedClub(    // member가 참여 불가능한 모임
+                savedMember,
+                List.of(savedPet),
+                Set.of(Gender.FEMALE),
+                Set.of(SizeType.SMALL)
+        ));
+        clubs.add(createSavedClub(    // member가 참여 가능한 모임
+                savedMember,
+                List.of(savedPet),
+                Set.of(Gender.values()),
+                Set.of(SizeType.values())
+        ));
+
+        Member member = memberRepository.save(new Member("member", "sdfdfef3", "imageUrl"));
+        Pet pet = petRepository.save(
+                new Pet(member, "뚱땡이", "수컷 대형견", LocalDate.now().minusDays(1L), SizeType.LARGE, Gender.MALE, "imageUrl"));
+
+        FindClubByFilterRequest request = new FindClubByFilterRequest(
+                FilterCondition.ABLE_TO_JOIN.name(),
+                province,
+                city,
+                village,
+                Arrays.stream(Gender.values()).map(gender -> gender.name()).collect(Collectors.toSet()),
+                Arrays.stream(SizeType.values()).map(sizeType -> sizeType.name()).collect(Collectors.toSet()),
+                20,
+                LocalDateTime.of(9999, 12, 31, 23, 59, 59),
+                999999999L
+        );
+
+        FindClubPageByFilterResponse responses = clubQueryService.findByFilter(member.getId(), request);
+        assertThat(responses.content()).hasSize(1);
     }
 
     @DisplayName("페이지 사이즈보다 더 많은 모임이 존재하는 경우, 한번에 페이지 사이즈 만큼의 모임이 조회된다.")

